### PR TITLE
feat: extension bisect

### PR DESCRIPTION
**Pull request by SychO9**
**Fixes #0000**

**Changes proposed in this pull request:**
https://discuss.flarum.org/d/32734-extension-bisect-utility

**Screenshot**
image
image
image
image
image
image
image
image

**Necessity**

- [ ] Has the problem that is being solved here been clearly explained?
- [ ] If applicable, have various options for solving this problem been considered?
- [ ] For core PRs, does this need to be in core, or could it be in an extension?
- [ ] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [ ] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

### CODE REVIEW

File: framework/core/js/src/admin/components/AdvancedPage.tsx [hunk 1]
```diff
@@ -8,9 +8,13 @@ import FormSectionGroup, { FormSection } from './FormSectionGroup';
 import ItemList from '../../common/utils/ItemList';
 import InfoTile from '../../common/components/InfoTile';
 import { MaintenanceMode } from '../../common/Application';
+import Button from '../../common/components/Button';
+import classList from '../../common/utils/classList';
+import ExtensionBisect from './ExtensionBisect';
 
 export default class AdvancedPage<CustomAttrs extends IPageAttrs = IPageAttrs> extends AdminPage<CustomAttrs> {
   searchDriverOptions: Record<string, Record<string, string>> = {};
+  modalThroughUrlShown = false;
```
Comment (askvortsov1) [inline framework/core/js/src/admin/components/AdvancedPage.tsx]:
```suggestion
  urlRequestedModalHasBeenShown = false;
```

File: framework/core/js/src/admin/components/AdvancedPage.tsx [hunk 2]
```diff
@@ -143,6 +153,18 @@ export default class AdvancedPage<CustomAttrs extends IPageAttrs = IPageAttrs> e
               <strong className="helpText">{app.translator.trans('core.admin.advanced.maintenance.options.' + app.data.maintenanceMode)}</strong>
             </div>
           ) : null}
+          <div className="Form-group">
+            <label>{app.translator.trans('core.admin.advanced.maintenance.bisect.label')}</label>
+            <p className="helpText">{app.translator.trans('core.admin.advanced.maintenance.bisect.help')}</p>
+            <Button
+              className={classList('Button', { 'Button--warning': app.data.bisecting })}
+              onclick={() => app.modal.show(ExtensionBisect)}
+              disabled={app.data.maintenanceMode && app.data.maintenanceMode !== MaintenanceMode.LOW_MAINTENANCE}
```
Comment (askvortsov1) [inline framework/core/js/src/admin/components/AdvancedPage.tsx:180]: I wish we had a way to show a "disabled tooltip"

File: framework/core/locale/core.yml [hunk 3]
```diff
@@ -11,6 +11,27 @@ core:
     advanced:
       description: "Configure advanced settings for your forum."
       maintenance:
+        bisect:
+          begin_button_text: Begin Bisect
+          continue_button_text: Continue Bisect
+          label: Extension Bisect
+          help: Helps to identify the extension causing an issue you observed. This automatically puts the forum in maintenance mode (low) until the process is over.
```
Comment (askvortsov1) [inline framework/core/locale/core.yml]:
```suggestion
          help: Helps to identify the extension causing some issue you observed. This automatically puts the forum in maintenance mode (low) until the process is over.
```

File: framework/core/locale/core.yml [hunk 4]
```diff
@@ -11,6 +11,27 @@ core:
...
+          description: >
+            This puts the forum in maintenance mode (low) until the process is over.
+            In each step some extensions will be disabled until we find the one causing the issue. So do not be surprised to see missing features or different theme design.
+            Keep this page open, and try to reproduce the issue after each step in a separate page. You can always come back to this page if you mistakenly close it.
+          end_button: Close
+          issue_question: Does the issue still occur?
+          issue_question_help: Try reproducing the issue in a separate page and answer based on the result.
+          no_button: No
+          result_description: >
+            Forum is no longer in maintenance mode. Extension bisect is over. Based on the information in each step, the cause of the issue is the following extension:
```
Comment (askvortsov1) [inline framework/core/locale/core.yml]:
```suggestion
            Forum is no longer in maintenance mode. Extension bisect is over. Based on your responses to each step, the cause of the issue is the following extension:
```

File: framework/core/src/Extension/Bisect.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+use Composer\Command\ClearCacheCommand;
+use Flarum\Settings\SettingsRepositoryInterface;
+use RuntimeException;
+use Symfony\Component\Console\Input\ArrayInput;
+use Symfony\Component\Console\Output\NullOutput;
+
+class Bisect
+{
+    protected BisectState $state;
+    protected bool $break = false;
```
Comment (askvortsov1) [inline framework/core/src/Extension/Bisect.php:27]: Could you add a comment explaining that `break` is present for when we will be running the bisecter across multiple instances of processes (i.e. requests), if that's a correct understanding of why we need it?

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
Should there be a way to stop the bisect halfway through?

**Comment by SychO9** [comment]
image
image

**Review by askvortsov1 — APPROVED** [review]
Awesome, very exciting!

**Pull request merged**